### DISCUSSION TITLE
How to capture accordion panel expansion in a custom component?

### DISCUSSION THREAD

**Discussion by alacatos** [mantinedev/mantine#6742] [Q&A] [answered]
Having an uncontrolled accordion, I wish to do some data fetching whenever an item is expanded. 

```js
const data = [] // fetched from server

return (
<Accordion>
  {data.map((item, i) => (
    <AccordionItem value={item.id}>
      <AccordionControl>{item.name}</AccordionControl>
      <AccordionPanel>
        <CustomPanel data={item}/> // should do data fetching when expanded
      </AccordionPanel>
    </AccordionItem>
  )})
  </Accordion>
)
```

`CustomPanel` does additional data fetching, my problem is that it fires the request as soon as the parent `Accordion` is mounted. I wish to do that on demand, whenever the item is expanded. It is not possible to fetch all the data before rendering the UI due to performance issues.

One workaround that came to my mind is to make the `Accordion` controlled, pass the control value down the tree to `CustomPanel` and then write some logic. I am not happy with this approach because the component tree is quite deep (4+ components), setting up a React context for this does not seem the best solution either.

What approach do you recommend?

I need this behavior in multiple places in the app

**Comment by rtivital** [comment, marked as answer]
Use controlled state, it is the correct solution
